src/axom/quest/IntersectionShaper.hpp

```diff
@@ -818,17 +1037,361 @@ class IntersectionShaper : public Shaper
+   * a volume fraction to a zone, the code adds the allowed amount to the zone.
+   * In most cases, this is just the computed material volume, though it can be
+   * restricted by amounts of other materials if replacing materials. Once a
+   * volume fraction has been added, a list of update materials is iterated and
+   * a corresponding amount is subtracted from them. This update list consists of
+   * "completely free", all non-shape  materials, and finally the shape material
+   * itself. The shape material is added at the end in case the initial volume
+   * fraction addition exceeded 1.
+   *
+   * The "completely free" material (CF) represents any volume fraction in the
+   * zones that has not been assigned to any material. We subtract from CF first
+   * as part of the update process so we do not have to subtract from real materials
+   * in the event that a zone is not full.
+   *
+   * The list of update materials depends on the shape's material replacement rule.
+   *
+   * Example:
+   *                        update mats
+   *                        |---->
+   *
+   *                  mat1  |  CF     mat0
+   *                  ---------------------
+   * add 0.4 mat1  -> | 0.4 | 0.2  | 0.4  |
+   *                  ---------------------
+   *
+   *                  mat1     CF     mat0     subtract
+   *                  ---------------------    -------
+   *                  | 0.8 | 0.2  | 0.4  |    | 0.4 |  (added 0.4 to mat1)
+   *                  ---------------------    -------
+   *
+   *                  mat1     CF     mat0     subtract
+   *                  ---------------------    -------
+   *                  | 0.8 | 0.0  | 0.4  |    | 0.2 |  (subtracted 0.2 from CF)
+   *                  ---------------------    -------
+   *
+   *                  mat1     CF     mat0     subtract
+   *                  ---------------------    -------
+   *                  | 0.8 | 0.0  | 0.2  |    | 0.0 |  (subtracted 0.2 from mat0)
+   *                  ---------------------    -------
+   */
+  template <typename ExecSpace>
+  void applyReplacementRulesImpl(const klee::Shape& shape)
+  {
+    // Make sure the material lists are up to date.
+    populateMaterials();
+
+    // Get this shape's material, creating the GridFunction if needed.
+    auto matVF = getMaterial(shape.getMaterial());
+    int dataSize = matVF.first->Size();
+
+    // Get this shape's array.
+    auto shapeVolFracName =
+      axom::fmt::format("shape_vol_frac_{}", shape.getName());
+    auto* shapeVolFrac = this->getDC()->GetField(shapeVolFracName);
+    SLIC_ASSERT(shapeVolFrac != nullptr);
+
+    // Allocate some memory for the replacement rule data arrays.
+    int execSpaceAllocatorID = axom::execution_space<ExecSpace>::allocatorID();
+    double* vf_subtract = axom::allocate<double>(dataSize, execSpaceAllocatorID);
+    double* vf_writable = axom::allocate<double>(dataSize, execSpaceAllocatorID);
```

Comment:
Would it make sense to use a `axom::Array` / `axom::ArrayView` here instead of `double *` ?

Reply:
I changed to using Array/ArrayView for these arrays. I had borrowed the allocate pattern from other methods in this class.